staging/src/k8s.io/apiserver/pkg/cel/openapi/resolver.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapi
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/kube-openapi/pkg/validation/spec"
+)
+
+type SchemaResolver interface {
+	ResolveSchema(gvk schema.GroupVersionKind) (*spec.Schema, error)
+}
+
+var ErrSchemaNotFound = fmt.Errorf("schema not found")
+
+// ClientDiscoveryResolver uses client-go discovery to resolve schemas at run time.
+type ClientDiscoveryResolver struct {
+	Discovery discovery.DiscoveryInterface
+}
+
+var _ SchemaResolver = (*ClientDiscoveryResolver)(nil)
+
+func (r *ClientDiscoveryResolver) ResolveSchema(gvk schema.GroupVersionKind) (*spec.Schema, error) {
+	p, err := r.Discovery.OpenAPIV3().Paths()
+	if err != nil {
+		return nil, err
+	}
+	resourcePath := resourcePathFromGV(gvk.GroupVersion())
+	c, ok := p[resourcePath]
+	if !ok {
+		return nil, fmt.Errorf("cannot resolve group version %q: %w", gvk.GroupVersion(), ErrSchemaNotFound)
+	}
+	b, err := c.Schema(runtime.ContentTypeJSON)
+	if err != nil {
+		return nil, err
+	}
+	resp := new(schemaResponse)
+	err = json.Unmarshal(b, resp)
+	if err != nil {
+		return nil, err
+	}
+	s, err := resolveType(resp, gvk)
+	if err != nil {
+		return nil, err
+	}
+	err = r.populateRefs(resp, s)
+	if err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+func (r *ClientDiscoveryResolver) populateRefs(resp *schemaResponse, schema *spec.Schema) error {
+	ref, isRef := refOf(schema)
+	if isRef {
+		// replace the whole schema with the referred one.
+		resolved, ok := resp.Components.Schemas[strings.TrimPrefix(ref, refPrefix)]
+		if !ok {
+			return fmt.Errorf("internal error: cannot resolve Ref %q: %w", ref, ErrSchemaNotFound)
+		}
+		*schema = *resolved
+	}
+	// schema is an object, populate its properties and additionalProperties
+	for name, prop := range schema.Properties {
+		err := r.populateRefs(resp, &prop)
+		if err != nil {
+			return err
+		}
+		schema.Properties[name] = prop
+	}
+	if schema.AdditionalProperties != nil && schema.AdditionalProperties.Schema != nil {
+		err := r.populateRefs(resp, schema.AdditionalProperties.Schema)
+		if err != nil {
+			return err
+		}
+	}
+	// schema is a list, populate its items
+	if schema.Items != nil && schema.Items.Schema != nil {
+		err := r.populateRefs(resp, schema.Items.Schema)
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
Handle or document the behavior if any of the other fields containing schemas (`oneOf`, `anyOf`, `not`, `patternProperties`, `dependencies`, `definitions`) are encountered? I don't expect us to ever use most of these, but oneOf might be used in the future for unions.

Reply:
Unfortunately the library handles only the subset that are observed to be used by `kube-openapi`. For `oneOf`, the only use cases would be `int-or-string` and `quantity`, which is correctly handled.

Reply:
> oneOf might be used in the future for unions.

I've wondered but not tried: can union be described in CEL?
```cel
[ has(self.a), has(self.b) ].exists_one(x, x)
```

Reply:
`exists_one` can be implemented with `filter` and `size`, something like this
`size([has(self.a), has(self.b)].filter(i, i)) == 1`

Reply:
Nevermind, we already have `exists_one`, your solution is cleaner